File: MergeSort.java
import java.util.Arrays;

class MergeSort {

    public static int[] mergeSort(int[] ls) {
        if(ls.length == 1) {
            return ls;
        }
        int len = ls.length;
        int[] left = Arrays.copyOfRange(ls, 0, len/2);
        int[] right = Arrays.copyOfRange(ls, len/2, len);
        left = mergeSort(left); // recursive step, break it down until array of length 1
        right = mergeSort(right); // recursive step, break it down until array of length 1

        int i = 0; // array index for left
        int j = 0; // array index for right
        int k = 0; // array index for sorted
        int[] sorted = new int[len];
        int[] leftOver;
        while(i < left.length && j < right.length) {
            if(left[i] < right[j]) {
                sorted[k] = left[i];
                i++;
            } else {
                sorted[k] = right[j];
                j++;
            }
            k++;
        }

        if(i < j) {
            leftOver = Arrays.copyOfRange(left, i, left.length);
        } else {
            leftOver = Arrays.copyOfRange(right, j, right.length);
        }

        for(int n = 0; n < leftOver.length; n++) {
            sorted[k] = leftOver[n];
            k++;
        }

        return sorted;
    }

    public static void main(String args[]) {
        // MergeSort test = new MergeSort();
        int[] arr = new int[]{1,2,3,4,5,6,7,8,9,10,11,12,12,11,10,9,8,7,6,5,4,3,2,1};
        System.out.println(Arrays.toString(arr));
        System.out.println(Arrays.toString(mergeSort(arr)));
    }

}

File: leetcode/merge_sort.py


def merge(ls1, ls2):
    ordered_list = []
    
    i = 0 # counter for ls1
    j = 0 # counter for ls2
    while(i < len(ls1) and j < len(ls2)):
        # adds values in combination list until one of the lists runs out
        if(ls1[i] < ls2[j]):
            ordered_list = ordered_list + [ls1[i]]
            i += 1
        else:
            ordered_list = ordered_list + [ls2[j]]
            j += 1
    
    if(i < len(ls1)):
        leftover = ls1[i:]
    else:
        leftover = ls2[j:]

    # leftover value is already sorted so we can add them together
    return ordered_list + leftover


def merge_sort(ls):
    if len(ls) == 1:
        # stops the recursion at list length 1
        # lists of length 1 are technically sorted
        return ls

    # splitting the list in half
    mid = len(ls)/2
    array_one = ls[:mid]
    array_two = ls[mid:]

    # recursive step to further break down the lists in half until it reaches a list of length 1
    array_one = merge_sort(array_one)
    array_two = merge_sort(array_two)

    # returns the merged list up one level
    return merge(array_one, array_two)


if __name__ == '__main__':
    import numpy as np
    x = np.arange(21)
    np.random.shuffle(x)
    print(x.tolist())
    print(merge_sort(x.tolist()))
